### ISSUE TITLE
[FEATURE PROPOSAL W/ IMPLEMENTATION] Faster SpriteBatch: Texture Array with LRU replacement strategy.

### ISSUE THREAD

**Issue by VaTTeRGeR** [libgdx/libgdx#5907]
There is a link to a working implementation for this suggestion right before the Benchmark section of this post.

#### Issue summary
The current SpriteBatch implementation doesn't make very good use of OpenGLs capabilities: It wastes performance by issuing a draw call (flushing) at every texture change (calling draw with a different texture), this can cause massive performance decreases when using a texture atlas consisting of more than one page.

I implemented an improved version of the SpriteBatch and Sprite to remove this limitation, based on a Tutorial be The Cherno: [Batch Rendering - Textures](https://www.youtube.com/watch?v=bw6JsLnx5Jg)

#### Reproduction steps/code
Render lot's of Sprites on a SpriteBatch with and without alternating Textures: [TexA, TexB, TexA, TexB...] vs [TexA, TexA, TexA, TexA, TexB...]. With the current SpriteBatch implementation the number of draw calls is proportional to the number of texture switches, dragging down the performance in games with lot's of textures in use in one scene. Yes, you could sort by texture before rendering, but in many games this isn't possible since you may need a specific draw order for your sprites.

#### Version of LibGDX and/or relevant dependencies
Up to and including: 1.9.10

#### Proposed Solution
Using a texture array instead of a single texture uniform in the fragment shader and passing the selected texture unit for each sprite as an vertex-attribute, this is surprisingly easy to implement! The Texture replacement should be done with the Least Recently Used (LRU) strategy to avoid constant switches once the texture units are all used up by other textures. All OpenGL features used in my implementation are GL20 compatible AFAIK.

#### Drawbacks
It requires some careful changes to one of the most central elements of LibGDX. Sprite vertices are now 24 bytes large instead of 20 bytes (one extra float). Old custom shaders will not work anymore with this unless you fall back to the old solution for custom shaders. I also don't really know how much refactoring is needed to integrate this into LibGDX neatly.

#### Reference Implementation
Since i use LibGDX for my game i implemented this optimization as a stripped down version of the SpriteBatch and Sprite class. You can use this as a reference in case you would like to add this optimization to LibGDX in the future.

All significant changes from the default Spritebatch are marked and commented with "//CHANGE: ..."
[My implementation: SpriteBatchMultitexture.java and SpriteMultitexture.java](https://github.com/VaTTeRGeR/orts_game/tree/43ba32a7707e8ae1484cc69b28e5e4bcf74d1839/game_src/com/badlogic/gdx/graphics/g2d)

#### Benchmarks
Here are some performance comparisons with 9 active Textures and different numbers of usable Texture Units, the first one acts like the old SpriteBatch and runs at 12 fps, mostly because my GPU (GTX960) can't handle the amount of draw calls:
![1_LRU_SpriteBatch](https://user-images.githubusercontent.com/5089550/73976416-8a896100-4928-11ea-85af-3e2e581e7c6d.jpg)

Even a small LRU Texture Cache (3 LRU) improves the performance a lot:
![3_LRU_SpriteBatch](https://user-images.githubusercontent.com/5089550/73976417-8bba8e00-4928-11ea-9a7b-cbbc99f83e77.jpg)

With more LRU slots than Textures (16 LRU for 9 Textures) it runs smoothly at 60 fps, most modern cards support 32 Texture Units, so you could have 32 independent textures without any significant slowdown compared to using a single texture:
![16_LRU_SpriteBatch](https://user-images.githubusercontent.com/5089550/73976418-8bba8e00-4928-11ea-8a56-789af4d188c4.jpg)

The effectiveness of this optimization depends on how often you switch textures within one drawing session of course and not so much on the amount of textures in use. What's shown above isn't a worst case demonstration, the benefit might be larger or smaller than this, depending on the use case.

**Comment by MobiDevelop** [comment]
I think you could avoid changing Sprite, by keeping the texture unit stuff only in your SpriteBatch. Also, if you can implement the Batch interface, that would make the impact minimal (or non-existent) on existing users - it's simply another option for people to use in place of the normal SpriteBatch.

I didn't look too closely at your changes but those were first things that come to my mind.

**Comment by VaTTeRGeR** [comment]
I agree, the more I look into it the better a custom Batch implementation looks to me. The only problem is the Sprite class, the Vertex attribute locations and sizes are hardcoded, the most promising workaround would be to let the optimized spritebatch inject the needed Vertex attributes while it copies the vertices. That would make the optimized Batch work with the unmodified Sprite class.

**Comment by Tom-Ski** [comment]
Interesting to see the performance on mobile for this. Have you ran any tests?

**Comment by VaTTeRGeR** [comment]
TL;DR Not yet. I can run a test on a weak Intel HD 3000 GPU afterwards though.

Mobile devices should benefit from the reduction of draw calls but might suffer slighly from the extra cpu overhead. I have the "#version 110" flag set in the fragment and vertex shader, so i hope the texture array also work on mobile. I can't test it on there yet though, since i do not have a mobile project nor Android export set up.

Since this is going to be a drop in replacement i would be very thankful if someone could paste the new batch class into his running project and give it a spin once it is ready (I still need to implement the other 11 draw methods...). It will work as a drop in replacement.

Here's a quick plot with "texture unit count" vs "draw calls" for ~15000 sprites with ~9 textures and different LRU sizes with linear and log scale:
![lru_linear_plot](https://user-images.githubusercontent.com/5089550/74034307-4f367300-49b8-11ea-86d0-2cb16d44bfc7.jpg)
![lru_log_plot](https://user-images.githubusercontent.com/5089550/74034310-4fcf0980-49b8-11ea-8d09-2ebcc54c6f8d.jpg)

That's a reduction from ~9300 to 7 draw calls. Draw calls get halved per LRU step.

**Comment by Tom-Ski** [comment]
It should work, but its a different gl implementation, gles vs gl, so they are not 100% directly comparable. They will benefit from reduced draw calls, no denying that. But there can be trade offs that effect mobile worse because of the mobile gpu compilers not being able to optimize. This could be similar to creating dependent texture reads. In a trivial case, the per-line cycle estimate goes from 4-10 using an array of textures where the index is coming from calculation.

**Comment by VaTTeRGeR** [comment]
Soooo... **i have a first fully compatible but mostly untested version of the optimized SpriteBatch ready to try out by anyone who is interested.** I tested it with my GUIs Stage but that's about it:

[-> TextureArraySpriteBatch.java](https://github.com/VaTTeRGeR/orts_game/blob/2038cf6b387c6bbef0ee60f68e47f79cc2405227/game_src/com/badlogic/gdx/graphics/g2d/TextureArraySpriteBatch.java)

The only incompatibility are custom shaders, they need to conform to the new attribute setup, the documentation is changed to reflect this though.

**Comment by VaTTeRGeR** [comment]
Lots and lots of small tweaks and fixes later - now it works on both Android and Desktop:

[TextureArraySpriteBatch.java](https://github.com/VaTTeRGeR/orts_game/blob/master/game_src/com/badlogic/gdx/graphics/g2d/TextureArraySpriteBatch.java)

**Comment by VaTTeRGeR** [comment]
**And finally, an Android benchmark! You can test how your phone handles it, the built APK is included under "/android/android-debug.apk".**

The demo shows 4 cases, i tried it on my 6yo Mi3: 
- ABABABAB pattern with SpriteBatch -> Slow (~250ms)
- **ABABABAB pattern with TextureArraySpriteBatch -> Fast (~7ms)**
- AAAAAAAA pattern with SpriteBatch -> Faster (~3ms)
- **AAAAAAAA pattern with TextureArraySpriteBatch -> Fast (~5ms)**

Source + APK: [libgdx-project.zip](https://github.com/libgdx/libgdx/files/4173267/libgdx-project.zip)


So far it looks good, the behavior is as expected :) SpriteBatch can't handle the draw calls when filling the screen in an ABABABAB pattern, while TextureArraySpriteBatch handles it nicely. Some percentage of performance is lost if you use an AAAAAAAAA pattern with the TextureArraySpriteBatch vs SpriteBatch, but i don't think it's significant compared to the benefits.

**Comment by xGnoSiSx** [comment]
@VaTTeRGeR 
This looks really good. Great Job!
I don't care that much for iOS - but if anyone with hardware and RoboVM experience could test this it would make it so much better.
I would love to have this as an option merged in libgdx or as a stand alone lib.

**Comment by Tom-Ski** [comment]
Awesome will take a look!

**Comment by VaTTeRGeR** [comment]
@xGnoSiSx 
Thanks!

If none of you find any bugs or errors in the documentation I'll open a pull request at the end of this week.

I reckon it's at least at a good level of stability. I did some rigorous bug-squishing, ironing out and optimizing over the weekend.

**Comment by piotr-j** [comment]
Ive tested on iOS and MacOS. MacOS seems to work just fine, as expected. On iOS there are few issues due to differences in `IntBuffer` implementation in RoboVM. Perhaps there is a smarter way to fix this, but heres what i come up with. RoboVM 

Issues Ive found (by it exploding). Generally casting directly to Buffer seems to fix the problem.

In constructor():
`Gdx.gl.glGetIntegerv(GL20.GL_MAX_TEXTURE_IMAGE_UNITS, (IntBuffer)texUnitsQueryBuffer.position(0));`
Im not certain setting position(0) does anything in this context as it is supposed to be a buffer, passing it in directly works.
It can also be moved up a line `((Buffer)texUnitsQueryBuffer).position(0);`.

Also
`((Buffer)textureUnitIndicesBuffer).flip();`

in flush():
```java
((Buffer)mesh.getIndicesBuffer()).position(0);
((Buffer)mesh.getIndicesBuffer()).limit(count);
```


Tested on devices with 11.4.1, 12.4.1, 14.2. Ive seen roughly 6-11x improvement in alternate test, more on slower devices as expected.

**Comment by tommyettinger** [comment]
I have also noticed the cast to Buffer working as a fix whenever I have a variant on SpriteBatch or some other NIO-buffer-using class in my code, and my code gets compiled by Java 9 or higher. There's also been some odd reports on the Discord of NIO Buffer-related code failing in similar ways, where `position(int)` isn't defined on `IntBuffer` or `ByteBuffer`. Does anyone else think we should preemptively change calls to `Buffer` methods to cast an `IntBuffer`/`ByteBuffer`/`ShortBuffer` to a `Buffer` before using, so if we ever have to change the build system to use newer Java, it won't suddenly break?

**Comment by anbu93** [comment]
This `ArrayTextureSpriteBatch` not working with spine?
And this not work on html, because GL30 not support.

**Comment by NathanSweet** [comment]
I dropped this into Spine for kicks. Spine does dashed line drawing using `GL_REPEAT` with the dash line image in its own texture. That's nice and easy, but showing many keyframes (an unusable amount) in Spine's graph makes for a lot of draw calls. This is what's being rendered (most of the dashed lines are hard to see because they overlap):
![](http://n4te.com/x/9880-tP1d.png)

SpriteBatch: 293,143 draw calls, 4 FPS
TextureArraySpriteBatch: 413 draw calls, 7 FPS

That's pretty extreme (25,460 keys) and it seems draw calls are not the only factor contributing to the slowdown.

Next, while rendering with TextureArraySpriteBatch I reduced the drawing until I just hit 144 FPS. Testing both:

SpriteBatch: 18,062 draw calls, 53 FPS
TextureArraySpriteBatch: 106 draw calls, 144 FPS

This is probably a better test, as it shows how much farther I can go with TextureArraySpriteBatch before performance drops. At 144 FPS I can draw 1,121 keys with SpriteBatch and 3,075 with TextureArraySpriteBatch, so ~3x. That's super cool! Especially since it's a drop-in replacement for SpriteBatch (for most people, I needed to change my shaders but it was easy). To deploy this to users I would fallback to SpriteBatch if `getMaxTextureUnits` is zero.

TL;DR: This seems useful, let's merge it! That will get more eyes on it in time. WDYT?

**Comment by tommyettinger** [comment]
I know a few places that could certainly benefit from faster access to multiple atlas pages! @BambooBandit I think has mentioned using either the associated PR #5914 or a different Batch to handle multi-page atlases, and he might have an opinion on changes that could be helpful for his usage. My own [TextraTypist](https://github.com/tommyettinger/textratypist) library frequently has to draw multiple fonts, with each font usually in its own Texture (especially if using FreeType), and drawing emoji also requires switching to a different Texture and back again. I'm interested in trying this out, especially if there's an easy enough fallback to SpriteBatch when GL30 isn't available (such as when using ANGLE).

**Comment by carlislefox** [comment]
One small comment, unsure if it is simply due to Fhox error or an oversight, but when I drop this into my existing project I hit an issue where the following exception is thrown:

```
Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: arraycopy: last destination index 24005 out of bounds for float[24000]
	at java.base/java.lang.System.arraycopy(Native Method)
	at com.badlogic.gdx.graphics.g2d.TextureArraySpriteBatch.draw(TextureArraySpriteBatch.java:683)
	at com.badlogic.gdx.graphics.g2d.NinePatch.draw(NinePatch.java:355)
	at com.badlogic.gdx.scenes.scene2d.utils.NinePatchDrawable.draw(NinePatchDrawable.java:49)
	at uk.co.harveydogs.mirage.engine.system.DrawModeSystem.renderSelectedBlueprint(DrawModeSystem.java:98)
	at uk.co.harveydogs.mirage.engine.system.DrawModeSystem.update(DrawModeSystem.java:79)
	at com.badlogic.ashley.core.Engine.update(Engine.java:240)
	at uk.co.harveydogs.mirage.ui.screen.RealmEditorScreen.render(RealmEditorScreen.java:71)
	at com.badlogic.gdx.Game.render(Game.java:48)
	at uk.co.harveydogs.mirage.RealmEditorGame.render(RealmEditorGame.java:43)
	at com.badlogic.gdx.backends.lwjgl3.Lwjgl3Window.update(Lwjgl3Window.java:387)
	at com.badlogic.gdx.backends.lwjgl3.Lwjgl3Application.loop(Lwjgl3Application.java:192)
	at com.badlogic.gdx.backends.lwjgl3.Lwjgl3Application.<init>(Lwjgl3Application.java:166)
	at uk.co.harveydogs.mirage.DesktopLauncher.launchRealmEditor(DesktopLauncher.java:119)
	at uk.co.harveydogs.mirage.DesktopLauncher.main(DesktopLauncher.java:38)
```

Which is the following line in draw():

```
            // Copy the vertices
            System.arraycopy(spriteVertices, srcPos, vertices, idx, spriteVertexSize);
```

If I pass in a size of 3000 in the constructor (instead of the default of 1000) then all is well, however with the default SpriteBatch I do not need to take this measure as I _guess_ it flushes before it hits the vertices limit? I'm out of my depth here this is too close to the hardware but just thought this worth mentioning before it gets merged as its a point of friction in an otherwise amazing drop in replacement.

As a side note, I am using this to instantiate my batch and it falls back flawlessly on ancient phones:

```
val batch by lazy {
        try {
            TextureArraySpriteBatch()
        } catch (e: Exception) {
            log.error(e) { "Unable to instantiate TextureArraySpriteBatch, falling back to default SpriteBatch implementation" }
            SpriteBatch()
        }
    }
```

Really excellent work dudes. Being able to leverage multiple textures and not worry about the render order enables me to do some awesome engine stuff I couldn't do before due to running out of space on a single atlas :D

UI  & Graphics on separate atlases with SpriteBatch (note binds & draws on the top right corner)

![Screenshot 2024-07-22 161404](https://github.com/user-attachments/assets/3ea8db5e-33a2-4c35-82f7-153a8e96ab00)

With TextureArraySpriteBatch

![Screenshot 2024-07-22 161100](https://github.com/user-attachments/assets/7dd2c9ff-4ba6-44b2-a68a-6d8a2be0b8ca)

Cheers
Fhox

**Comment by carlislefox** [comment]
Hey dudes me again,

Just wanted to say I've been working with my Macbook a lot lately and thought I'd test this out, I unfortunately hit the following:

`[main] ERROR uk.co.harveydogs.mirage.client.services.GraphicsService - Unable to instantiate TextureArraySpriteBatch, falling back to default SpriteBatch implementation
java.lang.IllegalStateException: Texture Arrays are not supported on this device:
Error compiling shader: Fragment shader:
ERROR: 0:1: '' :  version '150' is not supported

	at com.badlogic.gdx.graphics.g2d.TextureArraySpriteBatch.<init>(TextureArraySpriteBatch.java:129)
	at com.badlogic.gdx.graphics.g2d.TextureArraySpriteBatch.<init>(TextureArraySpriteBatch.java:105)
	at com.badlogic.gdx.graphics.g2d.TextureArraySpriteBatch.<init>(TextureArraySpriteBatch.java:99)
	at uk.co.harveydogs.mirage.client.services.GraphicsService$batch$2.invoke(GraphicsService.kt:14)
	at uk.co.harveydogs.mirage.client.services.GraphicsService$batch$2.invoke(GraphicsService.kt:12)
	at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:74)
	at uk.co.harveydogs.mirage.client.services.GraphicsService.getBatch(GraphicsService.kt:12)
	at uk.co.harveydogs.mirage.client.MirageApplication$batch$2.invoke(MirageApplication.kt:17)
	at uk.co.harveydogs.mirage.client.MirageApplication$batch$2.invoke(MirageApplication.kt:17)
	at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:74)
	at uk.co.harveydogs.mirage.client.MirageApplication.getBatch(MirageApplication.kt:17)
	at uk.co.harveydogs.mirage.client.MirageApplication.render(MirageApplication.kt:30)
	at com.badlogic.gdx.backends.lwjgl3.Lwjgl3Window.update(Lwjgl3Window.java:387)
	at com.badlogic.gdx.backends.lwjgl3.Lwjgl3Application.loop(Lwjgl3Application.java:193)
	at com.badlogic.gdx.backends.lwjgl3.Lwjgl3Application.<init>(Lwjgl3Application.java:167)
	at uk.co.harveydogs.mirage.client.lwjgl3.Lwjgl3Launcher.main(Lwjgl3Launcher.kt:33)
	at uk.co.harveydogs.mirage.client.lwjgl3.Lwjgl3Launcher.main(Lwjgl3Launcher.kt)`

I actually do remember having this issue with another implementation of Texture Arrays a year or so ago in the Discord and we were able to work around it to the extent where it would complain but still work but... I can't find that code at the moment or remember what on earth we did, sorry :(

**Comment by crykn** [comment]
@carlislefox the incompatibility of the default shaders with macOS & OpenGL 3+ is an ongoing problem, see #5960 (which could definitely use some people chiming in btw). The PR also mentions a few ways one can work around this in one's own game.
